### PR TITLE
FH-3157 Add integration test for redis & mongo

### PR DESCRIPTION
**Pull request by david-martin**

### CODE REVIEW

File: lib/sync/index.js [hunk 1]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
...
+ * Connect to mongodb & redis with the given connection urls.
+ * This should be called before `start()` is called
+ * 
+ * @param {string} mongoDBConnectionUrl A MongoDB connection uril in a format supported by the `mongodb` module
+ * @param {string} redisUrl A redis connection url in a format supported by the `redis` module
+ * @param {function} cb 
+ */
+function connect(mongoDBConnectionUrl, redisUrl, cb) {
+  if (arguments.length < 3) throw new Error('connect requires 3 arguments');
+
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 2]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
...
+/**
+ * Connect to mongodb & redis with the given connection urls.
+ * This should be called before `start()` is called
+ * 
+ * @param {string} mongoDBConnectionUrl A MongoDB connection uril in a format supported by the `mongodb` module
+ * @param {string} redisUrl A redis connection url in a format supported by the `redis` module
+ * @param {function} cb 
+ */
+function connect(mongoDBConnectionUrl, redisUrl, cb) {
+  if (arguments.length < 3) throw new Error('connect requires 3 arguments');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync/index.js [hunk 3]
```diff
@@ -17,26 +22,34 @@ function toJSON(dataset_id, returnData, cb) {
...
 }
 
+// Functions that are allowed to be invoked
+var invokeFunctions = ['sync', 'syncRecords', 'listCollisions', 'removeCollision', 'setLogLevel'];
+
+// Invoke the Sync Server.
 function invoke(dataset_id, params, callback) {
   syncUtil.doLog(dataset_id, 'info', 'invoke');
 
+  if (arguments.length < 3) throw new Error('invoke requires 3 arguments');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync/index.js [hunk 4]
```diff
@@ -45,6 +58,7 @@ function init(dataset_id, options, cb) {
   cb(null, {});
 }
 
+// Stop cloud data sync for the specified dataset_id.
 function stop(dataset_id, cb) {
   syncUtil.doLog(dataset_id, 'info', 'stop');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 5]
```diff
@@ -17,26 +22,34 @@ function toJSON(dataset_id, returnData, cb) {
...
+    return callback('invalid fn parameter provided in params "' + fn + '"', null);
   }
+  var fnHandler = module.exports[fn];
 
   return fnHandler(dataset_id, params, callback);
 }
 
+// Initialise cloud data sync service for specified dataset.
 function init(dataset_id, options, cb) {
   syncUtil.doLog(dataset_id, 'info', 'init');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 6]
```diff
@@ -136,49 +188,45 @@ function start(cb) {
       return callback();
     },
     function startSyncScheduler(callback) {
-      syncScheduler = new SyncScheduler();
+      syncScheduler = new SyncScheduler(redisClient);
```
Comment (fh-qube) [inline lib/sync/index.js]: MAJOR Add the "var" keyword to this declaration of "syncScheduler". [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2703)

File: lib/sync/index.js [hunk 7]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
   }
 }
 
-function sync(dataset_id, params, callback) {
+function sync(dataset_id, params, cb) {
   syncUtil.doLog(dataset_id, 'info', 'sync');
 
   // TODO
```
Comment (fh-qube) [inline lib/sync/index.js]: INFO Complete the task associated to this TODO comment. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)

File: lib/sync/index.js [hunk 8]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
...
+    return cb(null, mongoDbClient, redisClient);
+  });
+}
+
+/**
+ * Starts all sync queues, workers & the sync scheduler.
+ * This should only be called after `connect()`.
+ * If this is not explicitly called before clients send sync requests,
+ * it will be called when the first client sends a sync request
+ *
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 9]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
...
   // TODO
 
   return cb(null, {});
 }
 
+
+/**
+ * Connect to mongodb & redis with the given connection urls.
+ * This should be called before `start()` is called
+ *
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 10]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
...
 }
 
+
+/**
+ * Connect to mongodb & redis with the given connection urls.
+ * This should be called before `start()` is called
+ * 
+ * @param {string} mongoDBConnectionUrl A MongoDB connection uril in a format supported by the `mongodb` module
+ * @param {string} redisUrl A redis connection url in a format supported by the `redis` module
+ * @param {function} cb
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 11]
```diff
@@ -53,6 +67,7 @@ function stop(dataset_id, cb) {
   cb(null, {});
 }
 
+// Stop cloud data sync service for ALL datasets.
 function stopAll(cb) {
   syncUtil.doLog(syncUtil.SYNC_LOGGER, 'info', 'stopAll');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 12]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
...
+    },
+    function connectToRedis(callback) {
+      var redisOpts = {
+        url: redisUrl
+      };
+      var client = redis.createClient(redisOpts);
+      return callback(null, client);
+    }
+  ], function(err, results) {
+    if (err) return cb(err);
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync/index.js [hunk 13]
```diff
@@ -89,38 +106,73 @@ function setLogLevel(dataset_id, params, cb) {
...
   return cb(null, {});
 }
 
-/* Synchronise the individual records for a dataset */
-function syncRecords(dataset_id, params, callback) {
+// Synchronise the individual records for a dataset
+function syncRecords(dataset_id, params, cb) {
   syncUtil.doLog(dataset_id, 'info', 'syncRecords');
 
   // TODO
```
Comment (fh-qube) [inline lib/sync/index.js]: INFO Complete the task associated to this TODO comment. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)

**Comment by fh-qube** [comment]
SonarQube analysis reported 19 issues:
* MAJOR 1 major
* MINOR 10 minor
* INFO 8 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/39779680c7f4eff1681d2614f59dae7cb7809eb3/lib/sync/index.js#L56) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/39779680c7f4eff1681d2614f59dae7cb7809eb3/lib/sync/index.js#L65) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/39779680c7f4eff1681d2614f59dae7cb7809eb3/lib/sync/index.js#L74) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/39779680c7f4eff1681d2614f59dae7cb7809eb3/lib/sync/index.js#L92) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/39779680c7f4eff1681d2614f59dae7cb7809eb3/lib/sync/index.js#L20) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/39779680c7f4eff1681d2614f59dae7cb7809eb3/lib/sync/index.js#L82) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)

File: lib/sync/index.js [hunk 14]
```diff
@@ -121,24 +123,52 @@ function syncRecords(dataset_id, params, cb) {
...
+ * This should be called before `start()` is called
+ * 
+ * @param {string} mongoDBConnectionUrl A MongoDB connection uril in a format supported by the `mongodb` module
+ * @param {string} redisUrl A redis connection url in a format supported by the `redis` module
+ * @param {function} cb 
+ */
 function connect(mongoDBConnectionUrl, redisUrl, cb) {
-  async.parallel([
+  if (arguments.length < 3) throw new Error('connect requires 3 arguments');
+
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Remove the useless trailing whitespaces at the end of this line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: lib/sync/index.js [hunk 15]
```diff
@@ -121,24 +123,52 @@ function syncRecords(dataset_id, params, cb) {
...
+ * Connect to mongodb & redis with the given connection urls.
+ * This should be called before `start()` is called
+ * 
+ * @param {string} mongoDBConnectionUrl A MongoDB connection uril in a format supported by the `mongodb` module
+ * @param {string} redisUrl A redis connection url in a format supported by the `redis` module
+ * @param {function} cb 
+ */
 function connect(mongoDBConnectionUrl, redisUrl, cb) {
-  async.parallel([
+  if (arguments.length < 3) throw new Error('connect requires 3 arguments');
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

File: lib/sync/index.js [hunk 16]
```diff
@@ -121,24 +123,52 @@ function syncRecords(dataset_id, params, cb) {
...
     function connectToRedis(callback) {
-      var client = redis.createClient(redisUrl);
+      var redisOpts = {
+        url: redisUrl
+      };
+      var client = redis.createClient(redisOpts);
       return callback(null, client);
     }
   ], function(err, results) {
     if (err) return cb(err);
```
Comment (fh-qube) [inline lib/sync/index.js]: MINOR Reformat the code to have only one statement per line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)

**Comment by fh-qube** [comment]
SonarQube analysis reported 19 issues:
* MAJOR 1 major
* MINOR 10 minor
* INFO 8 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L92) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L82) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* MAJOR [Add the "var" keyword to this declaration of "syncScheduler".](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L191) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS2703)
* MINOR [Reformat the code to have only one statement per line.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L32) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AOneStatementPerLine)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L20) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L121) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L55) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L112) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* MINOR [Remove the useless trailing whitespaces at the end of this line.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L64) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingWhitespace)
* INFO [Complete the task associated to this TODO comment.](https://github.com/feedhenry/fh-mbaas-api/blob/473a370df554b7b006990d34a0c88765c33a3e32/lib/sync/index.js#L74) [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1135)
* ... 3 more

File: Gruntfile.js [hunk 17]
```diff
@@ -37,13 +39,26 @@ module.exports = function(grunt) {
...
+      }
+    },
 
     // These are the properties that grunt-fh-build will use
-
     unit: _.map(unit_args, makeUnits),
     unit_cover: _.map(unit_args, makeUnitCovers)
   });
 
+  grunt.registerTask('integration', function() {
```
Comment (wei-lee) [inline Gruntfile.js]: should we run something here?

File: lib/sync/index.js [hunk 18]
```diff
@@ -121,24 +123,52 @@ function syncRecords(dataset_id, params, cb) {
...
+    return cb(null, mongoDbClient, redisClient);
   });
 }
 
+/**
+ * Starts all sync queues, workers & the sync scheduler.
+ * This should only be called after `connect()`.
+ * If this is not explicitly called before clients send sync requests,
+ * it will be called when the first client sends a sync request
+ *
```
Comment (wei-lee) [inline lib/sync/index.js]: in this case, we can added a in memory flag to return early if `start` is already being called.

**Pull request merged**